stan/math/prim/mat/meta/operands_and_partials.hpp

```diff
@@ -15,14 +15,14 @@ namespace internal {
 /* This class will be used for both multivariate (nested container)
    operands_and_partials edges as well as for the univariate case.
  */
-template <int R, int C>
-class ops_partials_edge<double, Eigen::Matrix<double, R, C>> {
+template <typename T_op, typename T_part, int R, int C>
```

Comment:
The templates needed to be changed to allow for cases where the input simplex was a vector of ```fvar<var>```, and the input densities were an array of vectors of doubles (i.e. ```log_mix(Matrix<fvar<var>, Dynamic, 1>, std::vector<Matrix<double, Dynamic, 1>>```)

This particular template (```ops_partials_edge<double, Eigen::Matrix<double, R, C>>```) did not need to be changed, but I thought it was better to maintain consistency with the others

Reply:
The operands_and_partials in this file and directory are for primitive types, and this isn't just a code organizational concern. The `dx` methods won't work if you're hoping these template specializations will capture and deal with types containing `<fvar<var>` appropriately.

Reply:
If I'm understanding the situation correctly, we'll need to change the code in `fwd/mat/meta/operands_and_partials` to deal with these types of nested Eigen matrices of `fvar<Dx>`s.

Reply:
The changes to ```fwd/mat/meta/operands_and_partials``` are also included in [the pull](https://github.com/stan-dev/math/pull/751/files#diff-7f6700f663c921cc2ba73d3b5175ad10).

> The ```dx``` methods won't work if you're hoping these template specializations will capture and deal with types containing ```fvar<var>``` appropriately.

There's a very good chance that I'm not fully understanding this, but the tests with```fvar<var>``` inputs all return the correct values? In the ```mix``` tests [here](https://github.com/andrjohns/math/blob/5767f8d6392c6a0c2c0494051126b9a74ad865a3/test/unit/math/mix/mat/fun/log_mix_test.cpp#L16)

Reply:
Hmm, I see that you have one new specialization in the PR for `operands_and_partials` for `std::vector<std::vector<fvar<Dx>>`. I was saying these matrix specializations that you changed in `prim/mat` from `double` to `T_op` that were intended to deal with both matrices of doubles and matrices of fvars might need to be split out instead into two versions, though it was a very good instinct to try to unify the code. But we actually want a version that does return 0 for all inputs here in `prim/mat` because we don't have derivatives w.r.t. `double` operands. It's interesting that your tests pass, because none of these operands_and_partials types in `prim/mat` have `dx()` implemented in such a way as to return anything other than 0 for all inputs... So I'm pretty curious what's happening here. Perhaps another specialization ends up covering these cases, if we are getting the right derivatives and everything for Eigen types?

Could you add new tests for these Eigen types for `fwd/mat/operands_and_partials`? Looks like the one we have now just does a single `std::vector`. It's a bit ironic, because we don't use forward mode (at all; obviously we don't use the new functionality you're adding here), but we ask that all modes are tested in the hopes that forward mode is one day fully tested and we can use it for new forms of HMC and optimization.


EDIT: I just looked it up - we do actually have specializations in `fwd/mat` for Eigen types `fvar<Dx>`. Did you realize that and were trying to add something new that I'm still misunderstanding? I'm sorry about the additional confusion, if so. Here is where I'm looking: https://github.com/stan-dev/math/blob/develop/stan/math/fwd/mat/meta/operands_and_partials.hpp#L40

Reply:
Ah, I think that this is actually my misunderstanding. Should ```mix``` tests for functions that take two inputs be testing combinations of ```double``` and ```fvar<var>```?

i.e. should the mix tests include ```foo(fvar<var>, double)``` and ```foo(fvar<fvar<var>>, double)``` or just ```foo(fvar<var>,fvar<var>)``` and ```foo(fvar<fvar<var>>, fvar<fvar<var>>)```?

Reply:
Ideally, yeah.

Reply:
Ok, I'm starting to drill down to the problem.

Operating under the ```develop``` branch, if a function calls:

```
  fvar<var> a(2.0, 2.0);
  double b = 2.0;

  operands_and_partials<fvar<var>, double> o(a, b);
  o.edge1_.partials_[0] = 5.0;
  if (!is_constant_struct<double>::value)
    o.edge2_.partials_[0] = 5.0;
```

```o.edge2_``` is correctly instantiated under the ```prim/scal/operands_and_partials``` template, since it allows the operand (```0p```) and the partial (```ViewElt```) to be different types.

On the other hand, if a function calls:
```
  fvar<var> a(2.0, 2.0);

  Eigen::Matrix<fvar<var>, Eigen::Dynamic, 1> avec(3);
  avec << a, a, a;

  Eigen::Matrix<double, Eigen::Dynamic, 1> bvec(3);
  bvec << 2.0, 2.0, 2.0;

  Eigen::Matrix<double, Eigen::Dynamic, 1> dxvec(3);
  dxvec << 5.0, 5.0, 5.0;

  operands_and_partials<Eigen::Matrix<fvar<var>, Eigen::Dynamic, 1>, 
                        Eigen::Matrix<double, Eigen::Dynamic, 1>> o(avec, bvec);
  o.edge1_.partials_vec_[0] = dxvec;
  if (!is_constant_struct<Eigen::Matrix<double, Eigen::Dynamic, 1>>::value)
    o.edge2_.partials_vec_[0] = dxvec;
```

```o.edge2_``` does not get instantiated under the ```prim/mat/operands_and_partials``` template because the template expects both the operand and the partial to be of type double.

So the templates between ```prim/scal``` and ```prim/mat``` are actually different in the combinations of types that they allow.

Reply:
That failure in operands_and_partials makes sense to me - we shouldn't have partial derivatives for a matrix of doubles. The derivative with respect to a constant is always 0. I think the problem might lie in `is_constant_struct` not understanding that a Matrix of doubles is constant?

Reply:
Quickly reading the code, it seems like I would expect `is_constant_struct` to work there, and [the tests](https://github.com/stan-dev/math/blob/develop/test/unit/math/prim/mat/meta/is_constant_struct_test.cpp#L9) seem to confirm this... How is that code executing?

Reply:
Oh wait, sorry, I'm just waking up. The compiler still needs that to compile, even if it statically evaluates to false and is never called at runtime. It looks like in your PR, the correct `prim/mat/meta/operands_and_partials` templates with a `partials_vec_` member are no longer being instantiated appropriately. I suspect if you revert your changes to `prim/mat/meta/operands_and_partials.hpp` that this will start resolving to the correct template again.

Reply:
This occurs under develop as well, not just my PR

Reply:
Ah, okay. Can you add a test for the proposed new behavior in operands_and_partials (required anyway) and I can dig into it this weekend or next week? This would be a new test file in `prim/mat/meta/operands_and_partials_test.cpp` showcases the things that should now work and things that shouldn't work. Thanks!

Reply:
I had a few minutes and am investigating further, driven partially by curiosity, and I did not notice that for some reason operands_and_partials thinks that the ViewElt of an eigen matrix is a `var`. That is where the problem stems from, I believe.

Reply:
I think we actually need to change the code here: https://github.com/stan-dev/math/blob/develop/stan/math/fwd/scal/meta/operands_and_partials.hpp#L63 to not pass Dx but actually whatever the ViewElt would be for Op2, 3, etc.

I will look more into it later. Having ViewElt = var isn't right..

Reply:
Progress is good! Unfortunately your changes break the usage of ```partials_return_type```.

If changing the test case posted above to:
```
#include <stan/math/mix/mat.hpp>
#include <stan/math/prim/scal/meta/is_constant_struct.hpp>
#include <gtest/gtest.h>

TEST(MetaTraits, partials_type) {
  using stan::math::fvar;
  using stan::math::var;
  using stan::is_constant_struct;
  using stan::math::operands_and_partials;

  typedef typename stan::partials_return_type<Eigen::Matrix<fvar<var>, Eigen::Dynamic, 1>,
                                              Eigen::Matrix<double, Eigen::Dynamic, 1>>::type
      T_partials_return;

  fvar<var> a(2.0, 2.0);

  Eigen::Matrix<fvar<var>, Eigen::Dynamic, 1> avec(3);
  avec << a, a, a;

  Eigen::Matrix<double, Eigen::Dynamic, 1> bvec(3);
  bvec << 2.0, 2.0, 2.0;

  //Derivative types set through partials_return_type
  Eigen::Matrix<T_partials_return, Eigen::Dynamic, 1> dxvec(3);
  dxvec << 5.0, 5.0, 5.0;

  operands_and_partials<Eigen::Matrix<fvar<var>, Eigen::Dynamic, 1>, 
                        Eigen::Matrix<double, Eigen::Dynamic, 1>> o(avec, bvec);

  if (!is_constant_struct<Eigen::Matrix<fvar<var>, Eigen::Dynamic, 1>>::value)
    o.edge1_.partials_vec_[0] = dxvec;
  if (!is_constant_struct<Eigen::Matrix<double, Eigen::Dynamic, 1>>::value)
    o.edge2_.partials_vec_[0] = dxvec;

  o.build(-1.0);
}
```

Compilation fails with:
```
lib/eigen_3.3.3/Eigen/src/Core/functors/AssignmentFunctors.h:24:104: error: assigning to
      'double' from incompatible type 'const stan::math::var'
  ...void assignCoeff(DstScalar& a, const SrcScalar& b) const { a = b; }
```

Reply:
Looks like it's still edge2 causing the failure:
```
    o.edge2_.partials_vec_[0] = dxvec;
                                ^
1 error generated.
```

Reply:
Ah, thanks for the improved test. It has been a while since I thought about mixed mode - I was forgetting that 1) we are only ever using a certain level of autodiff type at once (e.g. `fvar<var>` or `fvar<fvar<var>>` but never both) and that 2) a mixed mode o&p user would have vectors of vars to fake assign to the edges. I think having these new test cases in the PR would have helped a lot. Sorry this took me so long, but I think your solution is great! Couple of asks to get it in shape - can you standardize on one of either ViewElt / Op or T_part / T_op? I think I prefer the former (because the partials type might actually be a vector or matrix of T_part) but I think it's more important to just have them consistent (across all of the operands_and_partials.hpp files splayed out across the different directories). 

Second ask - can you add these tests and others to the PR? I think we either strongly like or require unit tests on new behavior like this, and as we've seen the tests are super helpful not just for validating but communication. 

I also would really like to see the original tests with their original values in the PR, if you don't mind. It should be easy enough to just include them as additional tests, I think?

I'm working on trying to refactor the 3 implementations of log_mix so that they are just 1 or 2 different ones... Seems like it should be possible but I don't want to ask you to do it without being sure.

Reply:
Re: refactoring / generalizing - I think the best we can do is, for the row/column vector cases, additionally templating on `, int R, int C` and getting rid of the other version. Here is what that looks like: https://gist.github.com/seantalts/2185963296cde0e216250de3006525f3

Reply:
Not a problem at all! 
So the to-do list is:
* Replace ```T_op``` and ```T_part``` with ```0p``` and ```ViewElt```
* Add ```operands_and_partials``` tests for new functionality (```std::vector<std::vector<T>>```)
* Add tests to ```mix/mat/meta/operands_and_partials``` for mixed-mode with autodiff types (```fvar<var>``` and ```fvar<fvar<var>>```) and containers of doubles
* Re-add original ```log_mix``` tests that were changed
* Condense different implementations of ```log_mix```

Anything I've missed/misunderstood?

Reply:
Sounds good. That should be `Op` and not `0p` but I think that's what you meant.

I think bullet 2 and 3 can be the same thing, pretty much. The new o&p use-cases you are fixing in this PR are all mixed-mode, and hopefully the old tests will continue to test the existing use-cases sufficiently. If you see any gaps there, even if you find yourself wondering "what would happen if..." feel free to add a test for that case! In this way, tests can also serve as documentation.

And the last one, I'm not sure there's a performant & readable way to condense the `std::vector` inner container specialization with the Eigen inner container ones. So don't worry about that for now, we can always come back to it.

Thank you very much for bearing with me!